nntrainer/tensor/tensor.cpp

```diff
@@ -2148,1680 +734,401 @@ Tensor &Tensor::dot_batched_deriv_wrt_2(Tensor &m_deriv,
-    ret = *(data + idx);
+void Tensor::copy(const Tensor &from) {
+  /// @todo enable copy to non-contiguous tensor
+  if (!itensor->getContiguous()) {
+    throw std::runtime_error("Cannot copy non-contiguous tensor");
+  }
 
-  } else if (getDataType() == ml::train::TensorDim::DataType::FP16) {
+  if (from.size() != 0 && size() == from.size() &&
+      getDataType() == from.getDataType()) {
+    // if tensor size and data type match, copy data
+    itensor->copy(from);
+  } else {
+    // replace with a new tensor that are the same with the given tensor
+    if (from.getDataType() == ml::train::TensorDim::DataType::FP32) {
+      Tensor t = Tensor(from.getDim(), from.getData<float>());
+      swap(t, *this);
+    } else if (from.getDataType() == ml::train::TensorDim::DataType::FP16) {
 #ifdef ENABLE_FP16
-    const _FP16 *data = getData<_FP16>();
-
-    unsigned int idx = isamax(len, data, 1);
-    ret = *(data + idx);
+      Tensor t = Tensor(from.getDim(), from.getData<_FP16>());
+      swap(t, *this);
 #else
-    throw std::invalid_argument("Error: enable-fp16 is not enabled");
+      throw std::invalid_argument("Error: enable-fp16 is not enabled");
 #endif
+    }
   }
-  return ret;
 }
 
-Tensor &Tensor::normalization(Tensor &output) const {
-  if (output.empty())
-    output = Tensor(dim);
-
-  output.copy(*this);
-  output.normalization_i();
-
-  return output;
-}
-
-void Tensor::normalization_i() {
-  NNTR_THROW_IF(!contiguous, std::invalid_argument)
-    << getName() << " is not contiguous, cannot do normalization.";
+void Tensor::copyData(const Tensor &from) { itensor->copyData(from); }
 
-  if (getDataType() == ml::train::TensorDim::DataType::FP32) {
-    const float *data = getData();
-
-    auto bounds = std::minmax_element(data, data + size());
-    const float min = *bounds.first;
-    const float max = *bounds.second;
-
-    if (max == min) {
-      Tensor tmp = *this;
-      this->subtract_i(tmp);
-    } else {
-      this->subtract_i(min);
-      this->divide_i(max - min);
+void Tensor::copy_with_stride(const Tensor &from) {
+  if (itensor->getDim() == from.getDim()) {
+    // if the tensor dim matches, copy the data
+    for (unsigned int b = 0; b < batch(); ++b) {
+      for (unsigned int c = 0; c < channel(); ++c) {
+        for (unsigned int h = 0; h < height(); ++h) {
+          for (unsigned int w = 0; w < width(); ++w) {
+            setValue(b, c, h, w, from.getValue<float>(b, c, h, w));
+          }
+        }
+      }
     }
-  } else if (getDataType() == ml::train::TensorDim::DataType::FP16) {
+  } else {
+    // replace with a new tensor that has the same data as the given tensor
+    Tensor t = Tensor(from.getDim(), true);
+    for (unsigned int b = 0; b < t.batch(); ++b) {
+      for (unsigned int c = 0; c < t.channel(); ++c) {
+        for (unsigned int h = 0; h < t.height(); ++h) {
+          for (unsigned int w = 0; w < t.width(); ++w) {
+            if (getDataType() == ml::train::TensorDim::DataType::FP32) {
```

Comment:
Ditto.

Reply:
Same issue as `copy()`.

Reply:
It is minor, how about moving the if part to outside of the loop?

Reply:
this is resolved now. thanks for pointing it out!